keps/sig-storage/20200116-node-failover.md

```diff
@@ -0,0 +1,178 @@
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
+approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+In case of an instance being shutdown (i.e. not running) the control plane doesn't make the right
+assumptions to enable stateful workload to fail-over, This KEP introduces a flow to do so.
+
+This KEP is part of collection of KEPs that intends to enhance stateful workload management on top of Kubernetes.
+
+## Motivation
+
+Today kubernetes doesn't enable providers that are able to safely detach (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new CSI capability that if advertised would trigger volume detach if a node is partitioned from the control plane.
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Implement in-cluster logic to handle node/control plane partitioning 
+* enable detach for all the storage providers
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is partitioned without any intervention when it's safe to do so
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP plans to introduce the following capability to the CSI capability to the spec:
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+	// SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations, when set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking for detach safety
+	// +optional
+  SafeDetach *bool
+}
+
+```
+
+
+the kube-controller-manager, would need to ensure that eviction looks at the `CSIDriver` and when `.spec.SafeDetach` is set true, forcefully delete pods.
```

Comment:
Please clarify how a CSI driver can determine when it is safe to detach.

`CSIDriver` API object is created when the CSI driver is deployed.  So it is a per driver setting that does not change once it is deployed.  By making this setting at this level, it suggests that a CSI driver can determine whether it is safe to detach ahead of time and it is always true or false.  I don’t think that is always the case.  I think there are situations when it is safe to detach but there are also situations when it is not safe to detach.

We need to first figure out how a CSI driver can make those decisions and then communicate that information with Kubernetes.